src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
+      // memory.copy(dst, src, C)  ==>  store(dst, load(src))
+      if (auto* csize = memCopy->size->dynCast<Const>()) {
+        auto bytes = csize->value.geti32();
+
+        switch (bytes) {
+          case 0: {
+            if (!destHasSideEffects &&
+                !EffectAnalyzer(getPassOptions(), features, memCopy->source)
+                   .hasSideEffects()) {
+              return ExpressionManipulator::nop(curr);
+            }
+            break;
```

Comment:
It would be more robust to replace the `MemoryCopy` with a block containing the dropped source and destination. That way the optimization can be done whether or not the operands have side effects. If they don't they will be cleaned up separately later on.